File: core/tests/test_correlations.cpp

// AI_GENERATED: C++ correlation tests
#include <gtest/gtest.h>
#include "pipeline_sim/correlations.h"
#include <cmath>

using namespace pipeline_sim;

class CorrelationTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Create test fluid
        fluid_.oil_density = 850.0;
        fluid_.gas_density = 0.85;
        fluid_.water_density = 1025.0;
        fluid_.oil_viscosity = 0.01;
        fluid_.gas_viscosity = 1.8e-5;
        fluid_.water_viscosity = 0.001;
        
        // Create test pipe
        auto n1 = std::make_shared<Node>("n1", NodeType::SOURCE);
        auto n2 = std::make_shared<Node>("n2", NodeType::SINK);
        pipe_ = std::make_shared<Pipe>("test_pipe", n1, n2, 1000.0, 0.2);
    }
    
    FluidProperties fluid_;
    Ptr<Pipe> pipe_;
};

TEST_F(CorrelationTest, BeggsBrillSinglePhase) {
    // Single phase oil
    fluid_.oil_fraction = 1.0;
    fluid_.gas_fraction = 0.0;
    fluid_.water_fraction = 0.0;
    
    BeggsBrillCorrelation correlation;
    auto result = correlation.calculate(fluid_, *pipe_, 0.05, 50e5, 300.0);
    
    EXPECT_NEAR(result.liquid_holdup, 1.0, 0.01);
    EXPECT_GT(result.pressure_gradient, 0.0);
    EXPECT_EQ(result.flow_pattern, FlowPattern::SEGREGATED);
}

TEST_F(CorrelationTest, BeggsBrillTwoPhase) {
    // Two-phase flow
    fluid_.oil_fraction = 0.7;
    fluid_.gas_fraction = 0.3;
    fluid_.water_fraction = 0.0;
    
    BeggsBrillCorrelation correlation;
    auto result = correlation.calculate(fluid_, *pipe_, 0.05, 50e5, 300.0);
    
    EXPECT_GE(result.liquid_holdup, 0.7);
    EXPECT_LE(result.liquid_holdup, 1.0);
    EXPECT_GT(result.pressure_gradient, 0.0);
}

TEST_F(CorrelationTest, CorrelationFactory) {
    auto beggs_brill = CorrelationFactory::create("Beggs-Brill");
    ASSERT_NE(beggs_brill, nullptr);
    EXPECT_EQ(beggs_brill->name(), "Beggs-Brill");
    
    auto available = CorrelationFactory::available_correlations();
    EXPECT_GE(available.size(), 4);
    
    // Test unknown correlation
    EXPECT_THROW(CorrelationFactory::create("Unknown"), std::runtime_error);
}

TEST_F(CorrelationTest, FlowPatternTransitions) {
    BeggsBrillCorrelation correlation;
    
    // Test flow pattern transitions with varying gas fraction
    std::vector<Real> gas_fractions = {0.0, 0.1, 0.3, 0.5, 0.7, 0.9, 0.95};
    FlowPattern last_pattern = FlowPattern::SEGREGATED;
    
    for (Real gf : gas_fractions) {
        fluid_.gas_fraction = gf;
        fluid_.oil_fraction = (1.0 - gf) * 0.8;
        fluid_.water_fraction = (1.0 - gf) * 0.2;
        
        auto result = correlation.calculate(fluid_, *pipe_, 0.1, 50e5, 300.0);
        
        // Flow pattern should change with gas fraction
        if (gf > 0.5) {
            EXPECT_NE(result.flow_pattern, FlowPattern::SEGREGATED);
        }
        
        last_pattern = result.flow_pattern;
    }
}

File: python/tests/test_correlations.py
# ===== python/tests/test_correlations.py =====
# AI_GENERATED: Test flow correlations
import pytest
import numpy as np
from pipeline_sim.correlations import BeggsBrill, HagedornBrown
import pipeline_sim as ps


class TestBeggsBrill:
    """Test Beggs-Brill correlation"""
    
    def test_single_phase_liquid(self):
        """Test single-phase liquid flow"""
        # Create test conditions
        network = ps.Network()
        n1 = network.add_node("n1", ps.NodeType.SOURCE)
        n2 = network.add_node("n2", ps.NodeType.SINK)
        pipe = network.add_pipe("pipe", n1, n2, 1000, 0.2)
        
        fluid = ps.FluidProperties()
        fluid.oil_fraction = 1.0
        fluid.gas_fraction = 0.0
        fluid.water_fraction = 0.0
        
        # Calculate
        result = BeggsBrill.calculate(fluid, pipe, flow_rate=0.05)
        
        assert result.liquid_holdup == 1.0
        assert result.pressure_gradient > 0
        assert result.flow_pattern_name in ["Segregated", "Distributed"]
    
    def test_two_phase_flow(self):
        """Test two-phase flow patterns"""
        network = ps.Network()
        n1 = network.add_node("n1", ps.NodeType.SOURCE)
        n2 = network.add_node("n2", ps.NodeType.SINK)
        pipe = network.add_pipe("pipe", n1, n2, 1000, 0.2)
        
        fluid = ps.FluidProperties()
        
        # Test different gas fractions
        gas_fractions = [0.1, 0.3, 0.5, 0.7, 0.9]
        
        for gf in gas_fractions:
            fluid.gas_fraction = gf
            fluid.oil_fraction = 1 - gf
            fluid.water_fraction = 0.0
            
            result = BeggsBrill.calculate(fluid, pipe, flow_rate=0.05)
            
            # Sanity checks
            assert 0 <= result.liquid_holdup <= 1.0
            assert result.pressure_gradient >= 0
            assert result.flow_pattern in range(4)
    
    def test_inclined_flow(self):
        """Test effect of pipe inclination"""
        network = ps.Network()
        n1 = network.add_node("n1", ps.NodeType.SOURCE)
        n2 = network.add_node("n2", ps.NodeType.SINK)
        
        fluid = ps.FluidProperties()
        fluid.oil_fraction = 0.7
        fluid.gas_fraction = 0.3
        
        # Test different inclinations
        inclinations = [-np.pi/2, -np.pi/4, 0, np.pi/4, np.pi/2]
        pressure_gradients = []
        
        for angle in inclinations:
            pipe = network.add_pipe(f"pipe_{angle}", n1, n2, 1000, 0.2)
            pipe.inclination = angle
            
            result = BeggsBrill.calculate(fluid, pipe, flow_rate=0.05)
            pressure_gradients.append(result.pressure_gradient)
        
        # Upward flow should have higher pressure gradient
        assert pressure_gradients[4] > pressure_gradients[2]  # Vertical up > horizontal
        assert pressure_gradients[0] < pressure_gradients[2]  # Vertical down < horizontal


class TestMechanisticModel:
    """Test mechanistic flow model"""
    
    def test_flow_pattern_prediction(self):
        """Test flow pattern predictions match expected ranges"""
        # This would test the mechanistic model implementation
        pass
